### ISSUE TITLE
LDS push happens 28 minutes after first configuration update (Gateway/VirtualService) of a custom gateway

### ISSUE THREAD

**Issue by vadimeisenbergibm** [istio/istio#16916]
**Bug description**

While performing the steps in https://istio.io/docs/tasks/traffic-management/egress/wildcard-egress-hosts/#wildcard-configuration-for-arbitrary-domains (it deploys a custom egress gateway pod) and creates a Gateway/VirtualService for it. The problem is that it takes a long time for LDS to be pushed to the gateway.

Once the LDS push is performed, the following lines appear in the pilot log:
```
scvg222: 0 MB released
scvg222: inuse: 19, idle: 43, sys: 62, released: 40, consumed: 22 (MB)
2019-09-08T16:50:18.747217Z     info    transport: closing server transport due to maximum connection age.
2019-09-08T16:50:18.747349Z     info    transport: loopyWriter.run returning. connection error: desc = "transport is closing"
2019-09-08T16:50:18.747414Z     info    ads     ADS: "172.30.109.69:45608" router~172.30.109.69~myegressgateway-7f67b544fb-xf95z.istio-system~istio-system.svc.cluster.local-477 terminated rpc error: code = Canceled desc = context canceled
2019-09-08T16:50:18.747418Z     info    transport: http2Server.HandleStreams failed to read frame: read tcp 172.30.109.77:15010->172.30.109.69:45608: use of closed network\
 connection
2019-09-08T16:50:19.235866Z     info    ads     ADS:CDS: REQ 172.30.109.69:57570 router~172.30.109.69~myegressgateway-7f67b544fb-xf95z.istio-system~istio-system.svc.cluster.local-504 282.485µs version:2019-09-08T16:21:25Z/133
2019-09-08T16:50:19.236844Z     info    ads     CDS: PUSH for node:myegressgateway-7f67b544fb-xf95z.istio-system clusters:57 services:32 version:2019-09-08T16:21:25Z/133
2019-09-08T16:50:19.237583Z     info    ads     EDS: PUSH for node:myegressgateway-7f67b544fb-xf95z.istio-system clusters:55 endpoints:55 empty:[]
2019-09-08T16:50:19.237752Z     info    ads     LDS: PUSH for node:myegressgateway-7f67b544fb-xf95z.istio-system listeners:1
```

It seems that some connection is closed due to maximum connection age, `router~172.30.109.69~myegressgateway-7f67b544fb-xf95z.istio-system~istio-system.svc.cluster.local-477` is terminated, `router~172.30.109.69~myegressgateway-7f67b544fb-xf95z.istio-system~istio-system.svc.cluster.local-504` is started and LDS is pushed.

Can someone explain what is this router and what is the possible course of events that the log describes?

Before the LDS is pushed, `istioctl proxy-status` reports that LDS is synched:

```
NAME                                                                 CDS        LDS        EDS        RDS          PILOT                           VERSION
myegressgateway-7d565c8bf4-5z8ln.istio-system                        SYNCED     SYNCED     SYNCED     NOT SENT     istio-pilot-6f9f85987-ckncd     1.3.0-rc.2
```

```
Clusters Match
Listeners Match
Routes Match
```

**Version (include the output of `istioctl version --remote` and `kubectl version`)** 
client version: 1.3.0-rc.2
citadel version: 1.3.0-rc.2
egressgateway version: 1.3.0-rc.2
galley version: 1.3.0-rc.2
ingressgateway version: 1.3.0-rc.2
pilot version: 1.3.0-rc.2
policy version: 1.3.0-rc.2
sidecar-injector version: 1.3.0-rc.2
telemetry version: 1.3.0-rc.2

**Affected product area (please put an X in all that apply)**

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Comment by howardjohn** [comment]
I dont quite understand. When did you apply the config?

From the logs it looks like the connection to myegressgateway was closed, then it immediately reconnected and config was pushed?

Was there 28 minutes between applying the config and the connection was closed?

**Comment by vadimeisenbergibm** [comment]
@howardjohn I deploy a custom egress gateway pod, myegressgateway. Then I define a Gateway/VirtualService. Once I define them it takes more than 20 minutes for LDS to be pushed to the gateway pod. 

It seems there is some glitch in the protocol between the Envoy of myegressgateway and the pilot. Pilot does not update Envoy on the change in the configuration (the Gateway/VirtualService). Only after some timeout (probably 30 minutes), when the connection between the pilot and the Envoy is closed, and a new connection is reopened, pilot sends the update to the Envoy.

**Comment by vadimeisenbergibm** [comment]
@howardjohn The problem happens only for the first configuration update (Gateway/VirtualService defined). All other updates are pushed immediately. I have updated the title accordingly.

**Comment by howardjohn** [comment]
Did this happen once if consistently? What do pilot logs show when the
virtual service is applied?

If it's reproducible can you provide all the setup config here? or do I
just need to follow those docs

**Comment by vadimeisenbergibm** [comment]
Steps to reproduce the bug:

1.
```
cat <<EOF > ./sni-proxy.conf
user www-data;

events {
}

stream {
  log_format log_stream '\$remote_addr [\$time_local] \$protocol [\$ssl_preread_server_name]'
  '\$status \$bytes_sent \$bytes_received \$session_time';

  access_log /var/log/nginx/access.log log_stream;
  error_log  /var/log/nginx/error.log;

  # tcp forward proxy by SNI
  server {
    resolver 8.8.8.8 ipv6=off;
    listen       127.0.0.1:8443;
    proxy_pass   \$ssl_preread_server_name:443;
    ssl_preread  on;
  }
}
EOF
```
2.
```
kubectl create configmap myegress-sni-proxy-configmap -n istio-system --from-file=nginx.conf=./sni-proxy.conf
```
3.
```
cat <<EOF | helm template install/kubernetes/helm/istio/ --name istio-egressgateway-with-sni-proxy --namespace istio-system -x charts/gateways/templates/deployment.yaml -x charts/gateways/templates/service.yaml -x charts/gateways/templates/serviceaccount.yaml -x charts/gateways/templates/autoscale.yaml -x charts/gateways/templates/role.yaml -x charts/gateways/templates/rolebindings.yaml --set global.istioNamespace=istio-system -f - > ./myegressgateway.yaml
gateways:
  enabled: true
  istio-ingressgateway:
    enabled: false
  istio-egressgateway:
    enabled: false
  myegressgateway:
    enabled: true
    labels:
      app: myegressgateway
      istio: myegressgateway
    replicaCount: 1
    autoscaleMin: 1
    autoscaleMax: 5
    cpu:
      targetAverageUtilization: 80
    serviceAnnotations: {}
    type: ClusterIP
    ports:
      - port: 443
        name: https
    configVolumes:
      - name: sni-proxy-config
        configMapName: myegress-sni-proxy-configmap
    additionalContainers:
      - name: sni-proxy
        image: nginx
        volumeMounts:
        - name: sni-proxy-config
          mountPath: /etc/nginx
          readOnly: true
EOF
```
4.
```
kubectl apply -f ./myegressgateway.yaml
```
5.
```
kubectl get pod -l istio=myegressgateway -n istio-system
```
6. 
```
kubectl apply -f - <<EOF
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: myegressgateway
spec:
  selector:
    istio: myegressgateway
  servers:
  - port:
      number: 443
      name: tls-egress
      protocol: TLS
    hosts:
    - "*.wikipedia.org"
    tls:
      mode: MUTUAL
      serverCertificate: /etc/certs/cert-chain.pem
      privateKey: /etc/certs/key.pem
      caCertificates: /etc/certs/root-cert.pem
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: direct-wikipedia-through-egress-gateway
spec:
  hosts:
  - "*.wikipedia.org"
  gateways:
  - myegressgateway
  tcp:
  - match:
    - gateways:
      - myegressgateway
      port: 443
    route:
    - destination:
        host: sleep.default.svc.cluster.local # the hostname does not matter for this example
        port:
          number: 443
      weight: 100
EOF
```

7. Check the listeners. It should contain the 443, but it will take about 30 minutes until the lister for the port 443 will appear.

```
istioctl proxy-config listeners $(kubectl get pod -l istio=myegressgateway -n istio-system -o jsonpath='{.items[0].metadata.name}') -n istio-system
```

Cleanup:

1.
```
kubectl delete vs direct-wikipedia-through-egress-gateway
kubectl delete gw myegressgateway
```
2.
```
kubectl delete -f ./myegressgateway.yaml
```
3.
```
kubectl create configmap myegress-sni-proxy-configmap -n istio-system
```

**Comment by vadimeisenbergibm** [comment]
@howardjohn The bug is reproducible, I reduced the docs to the minimal configuration, please check https://github.com/istio/istio/issues/16916#issuecomment-529509129.

**Comment by vadimeisenbergibm** [comment]
Pilot logs:

myegressgateway is deployed:
```
2019-09-09T14:50:34.532242Z	info	Handle service myegressgateway in namespace istio-system
2019-09-09T14:50:34.543311Z	info	Handle EDS endpoint myegressgateway in namespace istio-system -> []
...
2019-09-09T14:50:34.681734Z	info	ads	EDS: PUSH for node:istio-ingressgateway-5549f44c4-nfcdf.istio-system clusters:112 endpoints:110 empty:[outbound_.443_._.myegressgateway.istio-system.svc.cluster.local outbound|443||myegressgateway.istio-system.svc.cluster.local]
2019-09-09T14:50:36.244250Z	info	ads	ADS:CDS: REQ 172.30.146.81:37160 router~172.30.146.81~myegressgateway-7d565c8bf4-pz8tz.istio-system~istio-system.svc.cluster.local-1903 222.637µs version:
2019-09-09T14:50:36.245074Z	info	ads	CDS: PUSH for node:myegressgateway-7d565c8bf4-pz8tz.istio-system clusters:58 services:33 version:2019-09-09T14:50:34Z/183
2019-09-09T14:50:36.332234Z	info	ads	EDS: PUSH for node:myegressgateway-7d565c8bf4-pz8tz.istio-system clusters:56 endpoints:55 empty:[outbound|443||myegressgateway.istio-system.svc.cluster.local]
2019-09-09T14:50:36.335067Z	info	ads	LDS: PUSH for node:myegressgateway-7d565c8bf4-pz8tz.istio-system listeners:0
2019-09-09T14:50:37.719484Z	info	Handling event update for pod myegressgateway-7d565c8bf4-pz8tz in namespace istio-system -> 172.30.146.81
2019-09-09T14:50:37.719653Z	info	ads	Pushing router~172.30.146.81~myegressgateway-7d565c8bf4-pz8tz.istio-system~istio-system.svc.cluster.local-1903
2019-09-09T14:50:37.720474Z	info	ads	CDS: PUSH for node:myegressgateway-7d565c8bf4-pz8tz.istio-system clusters:58 services:33 version:2019-09-09T14:50:34Z/183
2019-09-09T14:50:37.721198Z	info	ads	EDS: PUSH for node:myegressgateway-7d565c8bf4-pz8tz.istio-system clusters:56 endpoints:55 empty:[outbound|443||myegressgateway.istio-system.svc.cluster.local]
2019-09-09T14:50:37.721240Z	info	ads	LDS: PUSH for node:myegressgateway-7d565c8bf4-pz8tz.istio-system listeners:0
2019-09-09T14:50:37.723681Z	info	Handle EDS endpoint myegressgateway in namespace istio-system -> []
2019-09-09T14:50:37.968576Z	info	Handling event update for pod myegressgateway-7d565c8bf4-pz8tz in namespace istio-system -> 172.30.146.81
2019-09-09T14:50:37.972861Z	info	Handle EDS endpoint myegressgateway in namespace istio-system -> [172.30.146.81]
2019-09-09T14:50:37.972878Z	info	ads	Full push, new service myegressgateway.istio-system.svc.cluster.local
2019-09-09T14:50:37.972884Z	info	ads	Endpoint updating service account spiffe://cluster.local/ns/istio-system/sa/myegressgateway-service-account myegressgateway.istio-system.svc.cluster.local
2019-09-09T14:50:38.073089Z	info	ads	Push debounce stable[186] 1: 100.17445ms since last change, 100.173905ms since last push, full=true
2019-09-09T14:50:38.074074Z	info	ads	XDS: Pushing:2019-09-09T14:50:38Z/184 Services:33 ConnectedEndpoints:11
2019-09-09T14:50:38.075618Z	info	ads	Cluster init time 1.503888ms 2019-09-09T14:50:38Z/184
```

After the virtual service/gateway are defined, I see the following:
```
2019-09-09T14:51:00.129297Z	info	ads	Pushing router~172.30.146.81~myegressgateway-7d565c8bf4-pz8tz.istio-system~istio-system.svc.cluster.local-1903
2019-09-09T14:51:00.129426Z	info	ads	EDS: PUSH for node:sleep-78c47bbbbc-ph6j9.foo clusters:56 endpoints:56 empty:[]
2019-09-09T14:51:00.130044Z	info	ads	CDS: PUSH for node:myegressgateway-7d565c8bf4-pz8tz.istio-system clusters:58 services:33 version:2019-09-09T14:51:00Z/185
2019-09-09T14:51:00.130370Z	info	ads	LDS: PUSH for node:sleep-78c47bbbbc-z2cv8.bar listeners:44
2019-09-09T14:51:00.130714Z	info	ads	EDS: PUSH for node:myegressgateway-7d565c8bf4-pz8tz.istio-system clusters:56 endpoints:56 empty:[]
2019-09-09T14:51:00.130748Z	info	ads	LDS: PUSH for node:myegressgateway-7d565c8bf4-pz8tz.istio-system listeners:0
```

**Comment by duderino** [comment]
@vadimeisenbergibm does the problem also occur with 1.3.0-rc.1?

**Comment by howardjohn** [comment]
> 2019-09-09T14:51:00.130748Z	info	ads	LDS: PUSH for node:myegressgateway-7d565c8bf4-pz8tz.istio-system listeners:0

Interesting, so it does a push but generates no listeners.

> does the problem also occur with 1.3.0-rc.1?

I would say more importantly, does it on 1.2.5? As in, did we regress?

**Comment by vadimeisenbergibm** [comment]
@duderino let me check.

**Comment by duderino** [comment]
@howardjohn is right.  1.2.5 is the more important comparison.

**Comment by vadimeisenbergibm** [comment]
@duderino @howardjohn The problem does **not** exist on 1.2.5, exists on 1.3.0-rc.1.

**Comment by howardjohn** [comment]
Looking into it now. fyi @rlenglet potential regression

**Comment by vadimeisenbergibm** [comment]
@howardjohn By the end of the day, please update the issue with what you found. Let me continue to work on it tomorrow Israeli time.

**Comment by howardjohn** [comment]
@vadimeisenbergibm I think it may be related to protocol sniffing. You can test by setting env variable in pilot `PILOT_ENABLE_PROTOCOL_SNIFFING=false`. I am doing the same, seems to be the cause but need to make sure its not just a coincidence it got fixed when I did that (from pilot restarting as a result, or something)

**Comment by howardjohn** [comment]
@vadimeisenbergibm I thought I reproduced it once, but now I no longer can. Not sure why. Tearing down my whole env and trying again

**Comment by howardjohn** [comment]
Once it is broken
* restarting galley does not do anything
* removing the config and re applying does not do anything
* killing the gateway does not do anything.

Restarting Pilot *does* fix the problem permanantly

**Comment by howardjohn** [comment]
Maybe related to https://github.com/istio/istio/pull/16702, since I see `buildGatewayListeners: no gateways for router myegressgateway-7998f9cd5f-wgvsz.istio-system`

**Comment by howardjohn** [comment]
To reproduce it, remove the VS and GW, scale egress to 0 replicas, scale egress back to 1 replica, recreate the VS. No need to redploy pilot

**Comment by rshriram** [comment]
> Root cause is the gateway doesn't meet selector when it should -- proxy.WorkloadLabels is nil

yes if metadata is missing, then we fall back to k8s. But I thought @kyessenov  had a PR that fixed this. We were copying the metadata at each XDS request or something and Kuat pointed out that it was not the right thing to do per the specs..

**Comment by howardjohn** [comment]
Ok basically what is happening:

* Sidecars have a LABELS metadata.
* 3 weeks ago we started deriving workload labels from this: https://github.com/istio/istio/issues/16124
* Gateways do NOT have LABELS metadata, so they don't get these functionality
* In all the cases here when we don't send listeners, the gateway fails to get labels, since it reads from pods and we have  the standard kubernetes pod registration race condition
* Prior to https://github.com/istio/istio/pull/16702/files this probably wasn't that big of a deal, since we didn't read from WorkloadLabels. Instead, we read from service instances which would eventually be updated.

**Comment by rshriram** [comment]
and https://github.com/istio/istio/pull/16394

**Comment by infa-ddeore** [comment]
> Ok basically what is happening:
> 
> * Sidecars have a LABELS metadata.
> * 3 weeks ago we started deriving workload labels from this: #16124
> * Gateways do NOT have LABELS metadata, so they don't get these functionality
> * In all the cases here when we don't send listeners, the gateway fails to get labels, since it reads from pods and we have  the standard kubernetes pod registration race condition
> * Prior to https://github.com/istio/istio/pull/16702/files this probably wasn't that big of a deal, since we didn't read from WorkloadLabels. Instead, we read from service instances which would eventually be updated.

@howardjohn is there a fix to this issue? we are hitting similar problem where the config isnt pushed when HPA autoscaled the user ingress g/w

```
info	ads	LDS: PUSH for node:custom-ingressgateway-6f4c5d6ffb-jhtqc.default listeners:0
```

But after some time (5 mins to 40 mins), custom g/w are in sync

```
info	ads	LDS: PUSH for node:custom-ingressgateway-6f4c5d6ffb-jhtqc.default listeners:1
```

**Comment by howardjohn** [comment]
The above issue was fixed in 1.3, so its likely a different root cause. Opening a new issue with as much details as you have may help
